Add cohorts to courses

bodhish opened the pull request:
## Proposed Changes
- [x] New SPA ready Nav bar 
- [x] Re-build Students index and filter with new design
- [x] Update create students page 
- [x] Update edit students page  
- [x] Update bulk import page backend change 
- [x] create a new Cohort index with simple filter
- [x] create a new Cohort creation page 
- [x] Cohort edit page 
- [x] Create a team index page with filter 
- [x] team new page 
- [x] team edit page 
- [x] Add option to pick cohorts in courses index 
- [x] Add option to pick cohorts in exports 

# Pending 
- [x] when a student is moved from one cohort to another, disband his team 
- [x] Build cohort coach assignment  

@pupilfirst/developers

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [x] Check if route, query, or mutation authorization looks correct.
  - Add tests for authorization, if required.
- [x] Ensure that UI text is kept in I18n files.
- [ ] Update developer and product docs, where applicable.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Check if new tables or columns that have been added need to be handled in the following services:
  - `Users::DeleteAccountService`
  - `Courses::CloneService`
  - `Courses::DeleteService`
  - `Courses::DemoContentService`
  - `Levels::CloneService`
  - `Schools::DeleteService`
- [ ] Check if changes in _packaged_ components have been published to `npm`.
- [x] Add development seeds for new tables.
- [x] If the updates involve Graph mutations ensure that the files are migrated to the new approach without a mutator.


### Breaking Changes 
- `CreateStudentsMutator` will now accept `cohort_id` instead of `course_id`

## Pending 
- [x] moving strings to en.yml

## Before Merge 
- [x] Create an issue to update User documentation on docs.pupilfirst.com 
- [ ] Backup Database

harigopal commented on config/routes.rb:
```diff
@@ -71,12 +82,9 @@
     end
 
-    resources :courses, only: %i[index show new] do
+    resources :courses, only: %i[] do
```
Now that there are no _resourceful_ actions here, this might as well be a namespace:

```suggestion
    namespace :courses do
```

harigopal commented on app/javascript/layouts/school_router/components/SchoolRouter__Nav.res:
```diff
@@ -0,0 +1,252 @@
+%raw(`require("./SchoolRouter__Nav.css")`)
+
+let t = I18n.t(~scope="components.SchoolAdminNavbar__Root")
```
This translation scope for this component appears to be unchanged.

harigopal commented on db/migrate/20220425082348_add_cohorts.rb:
```diff
@@ -0,0 +1,142 @@
+      t.timestamps
+    end
+
+    add_reference :courses, :default_cohort, foreign_key: { to_table: :cohorts }
```
I had a discussion with @vinutv about this, and we think we might be able to avoid this _default_cohort_ selection.

Instead of assigning a default cohort, it should be sufficient to add a cohort selection on the form which converts an applicant into a student. There's a drawer right now that pops up asking for additional info right before an applicant is onboarded as a student. This form could be updated to include a cohort selection.

The other use of the default cohort that we discussed was a default selection in the _Add Students_ form - this can be safely ignored, I think.

bodhish replied:
This was supposed to be used for public signup. If you don't have a processing url, students would get onboarded automatically upon email verification which would need a default cohort.

harigopal replied:
Understood!

harigopal commented on db/schema.rb:
```diff
@@ -309,6 +337,12 @@
     t.boolean "dashboard_toured"
     t.integer "resume_file_id"
```
Could we use this opportunity to clean up some of these old fields? 😅

harigopal reviewed (changes requested):
I've only skimmed through the changes this time.

harigopal commented on app/javascript/shared/components/CourseResourcesFilter.res:
```diff
@@ -0,0 +1,183 @@
+  color: string,
+}
+
+let makeFilter = (key, label, filterType: filterType, color) => {
```
The type annotation `:filterType` shouldn't be required here.

bodhish commented on app/forms/cohorts/bulk_import_students_form.rb:
```diff
@@ -87,7 +87,7 @@ def strings_must_not_be_too_long
       end
 
-      errors.add(:base, 'One or more entries have invalid strings')
+      errors[:base] << 'One or more entries have invalid strings'
```
Unrelated change, should revert

harigopal reviewed (changes requested):
@bodhish  Results of today's review:

- Cohorts page filter should have a default filter selected: `Status: Active`. Replace `Include: Inactive Cohorts` with `Status: Ended`. This is needed because it's confusing when a person first visits a page, and the count at the bottom says _Only one cohort to show_ (I've updated the text for the pagination counter).
- The school admin _breadcrumbs_ UI appears to be useless. If we were to delete it, I don't think we'd miss out on any functionality. Everything except for the course selector, on the top-left of the secondary bar should probably be left out.
- The Students index page does not need a `Include: Inactive Students` filter any more. This is because _most_ "inactive" students will now be part of an ended cohort, and it's very weird when we pick an ended cohort, and the UI doesn't show any student by default, necessitating the addition of the `Include: Inactive Students` filter. Let's just not default to checking the status of the student on this list. This leaves the question of how to handle _dropped-out students_...
- ...which can probably be handled by finally adding a _Delete Student Profile_ option. After checking with @reenas, this feels like a cleaner option that having a _Dropped-out Student_ state, which is messier because of how it leaves an artefact that needs to be filtered out in a bunch of places, and needs custom handling on the student dashboard. We can probably leave this be for the time being.

bodhish commented:
@harigopal Updates 

1. I have updated the page title to fix the issue. 
<img width="922" alt="image" src="https://user-images.githubusercontent.com/14979190/191338326-5dcadb6a-ab6c-44f2-9149-320d94d1cfd9.png">
<img width="952" alt="image" src="https://user-images.githubusercontent.com/14979190/191338247-7cf34a21-f266-441c-b481-d941cc3d37e3.png">

2. Done, @vinutv can we increase the width for the course selector?
<img width="800" alt="image" src="https://user-images.githubusercontent.com/14979190/191339258-dd8815df-94c8-4286-bf9e-f868c77c1e02.png">

3. Done, the same query was getting reused in 2 other places; I have added an `active` filter to handle those cases https://github.com/pupilfirst/pupilfirst/commit/bed539de60deb8bdef2b5b9590f3904caf712ead

4. ✅

bodhish commented:
Clean up migration 

```
remove_column :founders, :startup_id
    remove_column :founders, :resume_file_id
    remove_column :founders, :dashboard_toured

    drop_table :faculty_course_enrollments
    drop_table :faculty_startup_enrollments
    drop_table :startups
    remove_column :courses, :ends_at

    # Clean up slack
    remove_column :faculty, :slack_username
    remove_column :faculty, :slack_user_id

    remove_column :founders, :slack_username
    remove_column :founders, :slack_user_id

    remove_column :targets, :slack_reminders_sent_at

    drop_table :public_slack_messages
    ```
